UTF-8 BOM stripping Redux

@sindresorhus @phated @contra @IsaacSanch

https://github.com/wearefractal/vinyl-fs/issues/72 this is awful. I understand your interest in making things generally better, and I like making tooling that fixes legacy problems as much as everyone else. 

But the contents of files are the user's domain. Not the tool's.

We wasted hours trying to find what in our big pipeline was changing our files. _Changing_ content of files is _not_ something that filesystem tools should do. To a filesystem, the contents of a file are _sacred_. Imagine if S3 or Dropbox or HTTP or even TCP subtly "tried to fix files for you". It would be a nightmare. **All cryptographically secure protocols break!**

I urge you to reconsider this. We were using vinyl and vinyl-fs for a bunch of ipfs pipelines, and would like to continue using it. We very much like the viewpoints described here: https://medium.com/@contrahacks/gulp-3828e8126466  But if the vinyl tooling _does not hold it as a **strong guarantee**_ that contents of files can move from A to B without magically changing underneath, we'll have to abandon it for something else. :'(

I think @contra said it best:

> Not adding it just because it was asked for, adding it because it makes sense. We shouldn't be fucking with contents without a way to disable it or we are a really shitty general purpose fs access tool (vinyl-fs)

And I agree. It was disappointing to find this design bug.

@contra i'm :+1: for this resolution, as it breaks no code (important!) even if i think the default is incorrect.

---

btw, i did not spend time looking for a non-utf8 file because:
- IMO i have found a test case: any file which is moved with vinyl and whose cryptographic checksum changes is (afaic) incorrect. 
- but if you want another answer, there are odd formats out there that simply just violate the expectations above. for example it is [actually quite common in scientific datasets](http://stackoverflow.com/questions/22229229/reading-4-byte-integers-from-binary-file-in-python) to have to have files that are _just_ a single array or matrix of 4 or 8 byte integers. In such cases, a file with a number whose hex value starts with `0xefbbbf` might match. one is provided in the repo below, with a single 4 byte number `0xefbbbf0a`.
- sure "no one will ever do this" -- well i found the problem by running into it :)

You can test both of these at https://github.com/jbenet/bom-diff-for-vinyl

```sh
> make test
./vmv.js
moved src/ to dst/
shasum src/test.num
34e399cb21690bfb4dbcc6308f96cd984461160b  src/test.num
shasum dst/test.num
adc83b19e793491b1c6ea0fd8b46cd9f32e592fc  dst/test.num
diff -r src dst
diff -r src/test.num dst/test.num
1c1
<
---
>
make: *** [test] Error 1
```

--- a/README.md
+++ b/README.md
@@ -50,6 +50,10 @@ fs.src(['*.js', '!b*.js'])
     - `false` will disable writing the file to disk via `.dest()`.
 
   - since - `Date` or `number` if you only want files that have been modified since the time specified.
+
+  - stripBOM - `true` or `false` if you want the BOM to be stripped on UTF-8 encoded files.
+    - Default value is `true`.
+
   - passthrough - `true` or `false` if you want a duplex stream which passes items through and emits globbed files.
     - Default is `false`.
 

--- a/lib/src/getContents/bufferFile.js
+++ b/lib/src/getContents/bufferFile.js
@@ -8,7 +8,13 @@ function bufferFile(file, opt, cb) {
     if (err) {
       return cb(err);
     }
-    file.contents = stripBom(data);
+
+    if (opt.stripBOM){
+      file.contents = stripBom(data);
+    } else {
+      file.contents = data;
+    }
+
     cb(null, file);
   });
 }

--- a/lib/src/getContents/streamFile.js
+++ b/lib/src/getContents/streamFile.js
@@ -4,8 +4,12 @@ var fs = require('graceful-fs');
 var stripBom = require('strip-bom-stream');
 
 function streamFile(file, opt, cb) {
-  file.contents = fs.createReadStream(file.path)
-    .pipe(stripBom());
+  file.contents = fs.createReadStream(file.path);
+
+  if (opt.stripBOM) {
+    file.contents = file.contents.pipe(stripBom());
+  }
+
   cb(null, file);
 }
 

--- a/lib/src/index.js
+++ b/lib/src/index.js
@@ -21,6 +21,7 @@ function src(glob, opt) {
   var options = assign({
     read: true,
     buffer: true,
+    stripBOM: true,
     sourcemaps: false,
     passthrough: false,
     followSymlinks: true

--- a/test/src.js
+++ b/test/src.js
@@ -121,7 +121,7 @@ describe('source stream', function() {
     stream.write(expectedFile);
   });
 
-  it('should strip BOM from UTF-8-encoded files', function(done) {
+  it('should strip BOM from UTF-8-encoded files by default', function(done) {
     var expectedPath = path.join(__dirname, './fixtures/bom-utf8.txt');
     var expectedContent = fs.readFileSync(expectedPath)
       // U+FEFF takes up 3 bytes in UTF-8: http://mothereff.in/utf-8#%EF%BB%BF
@@ -137,6 +137,26 @@ describe('source stream', function() {
     };
 
     var stream = vfs.src('./fixtures/bom-utf8.txt', {cwd: __dirname});
+
+    var buffered = [];
+    bufferStream = through.obj(dataWrap(buffered.push.bind(buffered)), onEnd);
+    stream.pipe(bufferStream);
+  });
+
+  it('should not strip BOM from UTF-8-encoded files if option is false', function(done) {
+    var expectedPath = path.join(__dirname, './fixtures/bom-utf8.txt');
+    var expectedContent = fs.readFileSync(expectedPath);
+
+    var onEnd = function(){
+      buffered.length.should.equal(1);
+      should.exist(buffered[0].stat);
+      buffered[0].path.should.equal(expectedPath);
+      buffered[0].isBuffer().should.equal(true);
+      bufEqual(buffered[0].contents, expectedContent).should.equal(true);
+      done();
+    };
+
+    var stream = vfs.src('./fixtures/bom-utf8.txt', {cwd: __dirname, stripBOM: false});
 
     var buffered = [];
     bufferStream = through.obj(dataWrap(buffered.push.bind(buffered)), onEnd);

